internal/cache/clockpro.go

```diff
@@ -398,13 +398,41 @@ func (c *shard) metaEvict(e *entry) {
 }
 
 func (c *shard) evict() {
-	for c.targetSize() <= c.sizeHot+c.sizeCold && c.handCold != nil {
-		c.runHandCold()
+	c.reclaim()
+	c.balance()
+}
+
+// Reclaim cache space by evicting cold entries. If the size of cold entries is
+// insufficient, then it continues with pushing handHot to demote hot entries
+// to cold entries.
+func (c *shard) reclaim() {
+	for c.targetSize() <= c.sizeHot+c.sizeCold {
+		if c.sizeCold > 0 {
+			c.runHandCold()
+		} else {
+			c.runHandHot()
+		}
+	}
+}
+
+// Balance hot/cold ratio according to the parameter c.coldTarget.
+func (c *shard) balance() {
+	// c.sizeHot > 0 is necessary because c.targetSize()-c.coldTarget can
+	// have negative values when cache memory is reserved. It prevents nil
+	// pointer exception when the list is empty but evict is called, and stops
+	// balancing when there are no hot entries in the cache list.
+	for c.targetSize()-c.coldTarget < c.sizeHot && c.sizeHot > 0 {
+		if c.handHot == c.handCold && c.sizeCold > 0 {
```

Comment:
Only pushing handCold when cold entries exist in the cache list. Because all cold entries are located between handCold and handHot, so handCold == handHot && c.sizeCold > 0 means handHot has caught up the handCold.